sql/ddl_log.cc

```diff
@@ -1112,6 +1113,24 @@ static void execute_rename_table(DDL_LOG_ENTRY *ddl_log_entry, handler *file,
                                     to_db->str, to_table->str, "",
                                     flags & FN_TO_IS_TMP);
   }
+
+  uint keys, total_keys;
+
+  if (!get_hlindex_keys_by_open(thd, from_db, from_table, from_path, &keys,
+                                &total_keys))
+  {
+    char idx_from[FN_REFLEN + 1], idx_to[FN_REFLEN + 1];
+    char *idx_from_end= strmov(idx_from, from_path);
+    char *idx_to_end= strmov(idx_to, to_path);
+
+    for (uint i= keys; i < total_keys; i++)
+    {
+      my_snprintf(idx_from_end, HLINDEX_BUF_LEN, HLINDEX_TEMPLATE, i);
+      my_snprintf(idx_to_end, HLINDEX_BUF_LEN, HLINDEX_TEMPLATE, i);
+      file->ha_rename_table(idx_from, idx_to);
+    }
+  }
+
```

Comment:
I'm not sure this is a good approach. It feels like you put too much inside one `DDL_RENAME_PHASE_TABLE` entry. There can be crashes between renames of individual vector indexes (presuming there will be many some day).

Perhaps, `DDL_RENAME_PHASE_TABLE` should be one `ha_rename_table` only? That is, a table with hlindexes would generate one `DDL_RENAME_PHASE_TABLE` entry **per index** (and one for the main table, of course)?

Reply:
Same for DROP, etc.

Reply:
> There can be crashes between renames of individual vector indexes (presuming there will be many some day).
Yes, sure. And my code handles it inline with what original code have been doing, that is ignores `non_existing_table_errors()`, e.g. if handler files were moved and .frm not. Or as you suggested above crash between different high-level indexes rename.

> ... a table with hlindexes would generate one DDL_RENAME_PHASE_TABLE entry *per index* (and one for the main table, of course)?
1. `PHASE` is just an `enum` member of an `ACTION`, it can't have stuff like table name or whatever attached.
2. `ACTION` is more logical (or high-level) entity, which may do many things like binlogging, handling stat tables, triggers, handler files and of course .frm.

Here're options:
1. We could add special phases to handle high-level indexes, but there's going to be no much difference between my solution. And it will introduce compatibility issues. Though it is very true that older versions won't be able to handle high-level indexes properly even with this patch.
2. We can't re-use existing actions like `DDL_LOG_RENAME_TABLE_ACTION`, as I mentioned before they're way to sophisticated for high-level indexes.
3. In theory we could try re-using special actions like `DDL_LOG_RENAME_ACTION` for high level indexes. That is issue one `DDL_LOG_RENAME_TABLE_ACTION` and then `DDL_LOG_RENAME_ACTION` for each high-level index. It may work, it has no compatibility concerns, but it is a question if we're going to be able to integrate it with e.g. `ALTER`.

Your thoughts?

Reply:
I don't think I'm much of an expert in ddl logging. Let's use your approach if it works.